File: app/models/goal.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from sqlalchemy import Date, DateTime, ForeignKey, Numeric, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class GoalStatus(str, Enum):
    active = "active"
    achieved = "achieved"
    archived = "archived"


class Goal(Base):
    __tablename__ = "goals"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    target_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False)
    deadline: Mapped[date | None] = mapped_column(Date(), nullable=True)
    status: Mapped[str] = mapped_column(
        String(16),
        nullable=False,
        default=GoalStatus.active.value,
        server_default=GoalStatus.active.value,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    user = relationship("User")
    transactions = relationship("Transaction", back_populates="goal", foreign_keys="Transaction.goal_id")

File: frontend/types/goal.ts
export type GoalStatus = 'active' | 'achieved' | 'archived';
export type GoalSystemKey = 'safety_buffer';

export type Goal = {
  id: number;
  user_id: number;
  name: string;
  target_amount: number;
  deadline: string | null;
  status: GoalStatus;
  is_system: boolean;
  system_key: GoalSystemKey | null;
  category_id: number | null;
  created_at: string;
  updated_at: string;
};

export type GoalWithProgress = Goal & {
  saved: number;
  percent: number;
  remaining: number;
  monthly_needed: number | null;
};

export type CreateGoalPayload = {
  name: string;
  target_amount: number;
  deadline?: string | null;
  category_id?: number | null;
};

export type UpdateGoalPayload = {
  name?: string;
  target_amount?: number;
  deadline?: string | null;
  category_id?: number | null;
};
